weave-js/src/components/Panel2/panelTree.ts

```diff
@@ -678,6 +679,99 @@ export const updateExpressionVarTypes = (node: PanelTreeNode, stack: Stack) => {
   });
 };
 
+export const updateExpressionVarNamesFromConfig = (
+  oldConfig: PanelTreeNode,
+  newConfig: PanelTreeNode
+) => {
+  const addedDelta = difference(oldConfig, newConfig);
+  const deletedDelta = difference(newConfig, oldConfig);
+  const path = getPathFromDelta(addedDelta);
+  const deletedPath = getPathFromDelta(deletedDelta);
+
+  const oldName = deletedPath[deletedPath.length - 1];
+  const newName = path[path.length - 1];
+
+  if (
+    // If the paths are the same, and the configs are the same except for the varName, we rename
+    path.slice(0, path.length - 1).join('.') ===
+      deletedPath.slice(0, deletedPath.length - 1).join('.') &&
+    _.isEqual(
+      getConfigForPath(getFullChildPanel(newConfig), path),
+      getConfigForPath(getFullChildPanel(oldConfig), deletedPath)
+    )
+  ) {
+    return updateExpressionVarNames(
+      newConfig,
+      [],
+      deletedPath,
+      oldName,
+      newName
+    );
+  }
+  return getFullChildPanel(newConfig);
+};
+
+export const updateExpressionVarNames = (
+  node: PanelTreeNode,
+  stack: Stack,
+  path: string[],
+  oldName: string,
+  newName: string
+) => {
+  return mapPanels(
+    node,
+    stack,
+    (child, childStack) => {
+      let newInputNode = child.input_node;
+      // marks null vars as dirty as well
+      // vars under the oldname will be null
+      const res = dereferenceAllVars(child.input_node, childStack, true);
+      for (const def of res.usedStack) {
+        if ((def as any).dirty) {
+          // if any of those variables are dirty, update the input_node
+          newInputNode = updateVarNames(
+            child.input_node,
+            childStack,
+            oldName,
+            newName
+          ) as NodeOrVoidNode;
+          break;
```

Comment:
is the operating assumption that we will only every have 1 dirty var and it is guaranteed to be the name change?

Reply:
if any var is dirty it updates the whole input_node, so it does not matter if theres 1+

Reply:
is it possible to have variables dirty for any other reason?

Reply:
just if they match the path, we sent in, so no, but if one was marked incorrectly, it wouldnt make too much of a difference unless the inputnode also had a variable that matched the one we are renaming.